config.org
* Notarock's Doom-Emacs Literate Configuration

This file is heavily inspired by doom-emacs's base `config.el` file. In fact,
it's just my personnal `config.el` translated to org-mode for your ready pleasures.

** User config
These are used for a number of things, particularly for GPG configuration,
some email clients, file templates and snippets.

#+BEGIN_SRC emacs-lisp
(setq user-full-name "Roch D'Amour"
      user-mail-address "roch.damour@gmail.com")
#+END_SRC

** Theming

THis function is probably the most important peice of configuration in this
whole repository.
#+BEGIN_SRC emacs-lisp
(setq doom-theme 'base16-brewer)
#+END_SRC

*** Font
Doom exposes five (optional) variables for controlling fonts in Doom. Here
are the three important ones:

+ `doom-font'
+ `doom-variable-pitch-font'
+ `doom-big-font' -- used for `doom-big-font-mode'

They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
font string.
#+BEGIN_SRC emacs-lisp
(setq font-family "Essential PragmataPro")
#+END_SRC

This snipper adjust the font size based on if I am currently using emacs on my
HiDPI laptop screen, or on anything else.

#+BEGIN_SRC emacs-lisp
(setq doom-font (font-spec :family font-family :size 22)
    doom-big-font (font-spec :family font-family :size 34))
#+END_SRC

want to change the style of line numbers, change this to `relative' or
`nil' to disable it:
#+BEGIN_SRC emacs-lisp
(setq display-line-numbers-type t)
#+END_SRC

*** Splash screen picture

Set this picture as the splash image

#+CAPTION: Notarock's splash image
#+NAME:   notarock.png
[[./notarock.png]]

#+BEGIN_SRC emacs-lisp
(setq fancy-splash-image "~/.doom.d/notarock.png")
#+END_SRC

Enable pixelwise resizing

#+BEGIN_SRC emacs-lisp
(setq frame-resize-pixelwise t)
#+END_SRC

*** Aesthetic modes

#+BEGIN_SRC emacs-lisp
; (feebleline-mode +1)
(dimmer-configure-magit)
(dimmer-configure-org)
(dimmer-mode t)
(setq highlight-indent-guides-method 'fill)
(add-hook 'prog-mode-hook 'highlight-indent-guides-mode)
#+END_SRC

Only show trailing whitespaces on programming, org and markdown mode.

#+BEGIN_SRC emacs-lisp
(add-hook 'prog-mode-hook (lambda ()
                             (setq show-trailing-whitespace t)))
(add-hook 'org-mode-hook (lambda ()
                             (setq show-trailing-whitespace t)))
(add-hook 'markdown-mode-hook (lambda ()
                             (setq show-trailing-whitespace t)))
#+END_SRC

** Aditionnal functions

This function reposition Emacs on startup:
- Width and height takes up 80% of the screen.
- Emacs is positionned right in the certer of the.

#+BEGIN_SRC emacs-lisp
(defun my/set-initial-frame ()
  "Set initial frame size and position"
  (let* ((base-factor 0.80)
         (a-width (* (display-pixel-width) (/ base-factor 2)))
         (a-height (* (display-pixel-height) base-factor))
         (a-left (truncate (/ (- (display-pixel-width) a-width) 2)))
         (a-top (truncate (/ (- (display-pixel-height) a-height) 2))))
    (set-frame-position (selected-frame) a-left a-top)
    (set-frame-size (selected-frame) (truncate a-width)  (truncate a-height) t)))
#+END_SRC

Launch the function.

#+BEGIN_SRC emacs-lisp
(my/set-initial-frame)
#+END_SRC

Top to bottom file indentation.

#+BEGIN_SRC emacs-lisp
(defun indent-buffer ()
  "Indent the whole buffer"
  (interactive)
  (save-excursion
    (indent-region (point-min) (point-max) nil)))
#+END_SRC

Insert a generated 64 character long random string.

#+BEGIN_SRC emacs-lisp
(defun insert-random-hash ()
  (interactive)
  (insert (string-trim (shell-command-to-string "< /dev/urandom tr -dc _A-Z-a-z-0-9 | head -c${1:-64};echo;"))))
#+END_SRC

`:q` should kill the current buffer rather than quitting emacs entirely
Need to type out :quit to close emacs

#+BEGIN_SRC emacs-lisp
(evil-ex-define-cmd "q" 'kill-this-buffer)
(evil-ex-define-cmd "quit" 'evil-quit)
#+END_SRC

** Langage-specific
*** Python

Python language server stupidity
#+begin_src emacs-lisp
(setq lsp-python-ms-executable (executable-find "python-language-server"))
#+end_src

*** Typescript
#+BEGIN_SRC emacs-lisp
;; (add-to-list 'auto-mode-alist '("\\.ts\\'" . typescript-mode))
;; (add-to-list 'auto-mode-alist '("\\.tsx\\'" . typescript-mode))
;; (after! flycheck
  ;; (flycheck-add-mode 'javascript-eslint 'typescript-mode)
  ;; (flycheck-add-mode 'css-stylelint 'typescript-mode)
  ;; (add-hook 'typescript-mode-hook (lambda () (flycheck-add-next-checker 'lsp-ui 'javascript-eslint)))
  ;; (add-hook 'typescript-mode-hook (lambda () (flycheck-add-next-checker 'javascript-eslint 'css-stylelint))))

(add-hook 'js2-mode-hook 'prettier-js-mode)
(add-hook 'web-mode-hook 'prettier-js-mode)

(map! :map web-mode-map
      :n "SPC m F" #'eslint-fix)

(map! :map typescript-mode-map
      :n "SPC m F" #'eslint-fix)
#+END_SRC

** Mode-specific
*** Org-mode
Contains All org-mode related configuration

Good looking bullet point, all about the eye-candy
#+BEGIN_SRC emacs-lisp
(use-package org-fancy-priorities
  :hook (org-mode . org-fancy-priorities-mode)
  :config
  (setq org-fancy-priorities-list '("■" "■" "■")))
#+END_SRC

Org-mode configurations

#+begin_src emacs-lisp
(after! org
  (map! :map org-mode-map
        :n "m-j" #'org-metadown
        :n "m-k" #'org-metaup)
  (setq ispell-local-dictionary "fr-toutesvariantes"
        org-directory "~/org/"
        org-journal-file (concat org-directory "journal.org")
        org-todos-file (concat org-directory "todos.org")
        org-capture-templates       (doct '(("Journal" :keys "j"
                                             :file org-journal-file
                                             :datetree t
                                             :todo nil
                                             :template ("* %^{description}"
                                                        ":properties:"
                                                        ":created: %u"
                                                        ":end:"
                                                        "%?"))
                                            ("Todos" :keys "t"
                                             :file org-todos-file
                                             :todo-state "TODO"
                                             :template ("* TODO: %^{description}"
                                                        ":properties:"
                                                        ":created: %u"
                                                        ":end:"))
                                            ))
        org-todo-keyword-faces (quote (("todo" :foreground "#ff6347" :weight bold)
                                       ("done" :foreground "#006400" :weight bold :strike-through t)))
        org-todo-keywords '((sequence "todo(t)" "done(d)"))
        org-log-done t))
#+end_src

*** Markdown-mode
#+begin_src emacs-lisp
(after! markdown
  (setq ispell-local-dictionary "fr-toutesvariantes"))
#+end_src

** Keybinds
*** Text manipulation
#+BEGIN_SRC emacs-lisp
(map! :ne "C-S-k" #'drag-stuff-up)
(map! :ne "C-S-j" #'drag-stuff-down)
(map! :ne "C-S-l" #'drag-stuff-right)
(map! :ne "C-S-h" #'drag-stuff-left)
#+END_SRC
*** Comment, indents, etc
#+BEGIN_SRC emacs-lisp
(map! :ne "SPC =" #'indent-buffer)
(map! :ne "SPC #" #'comment-or-uncomment-region)
#+END_SRC
*** Project navigation
Uses dumb-jump to find functions, variables, and other definition
#+BEGIN_SRC emacs-lisp
(map! :ne "SPC j g" #'dumb-jump-go)
(map! :ne "SPC j b" #'dumb-jump-back)
#+END_SRC
*** Text insertion

Binds random hash to "space [i]nsert [h]ash"

#+BEGIN_SRC emacs-lisp
(map! :ne "SPC i h" #'insert-random-hash)
#+END_SRC

*** Windows & frame manipulation

Split the current window and focus the newly created frame

#+BEGIN_SRC emacs-lisp
(map! :ne "SPC w V" (lambda () (interactive)(evil-window-vsplit) (other-window 1)))
#+END_SRC

** Others
*** Hacks
Org capture weird behaviour fix

#+BEGIN_SRC emacs-lisp
(map! [remap org-capture] nil)
#+END_SRC

Function used to recompile this repository's README

#+BEGIN_SRC emacs-lisp
;;  (shell-command "pandoc config.org -o README.md")
#+END_SRC

*** modes
Enable global git-gutter-mode

#+BEGIN_SRC emacs-lisp
(global-git-gutter-mode +1)
#+END_SRC

Enable global wakatime-mode
#+BEGIN_SRC emacs-lisp
(global-wakatime-mode +1)
#+END_SRC

*** Hooks
#+begin_src emacs-lisp
(setq ispell-dictionary "en_US")
#+end_src


config.el
(setq user-full-name "Roch D'Amour"
      user-mail-address "roch.damour@gmail.com")

(setq doom-theme 'base16-brewer)

(setq font-family "Essential PragmataPro")

(setq doom-font (font-spec :family font-family :size 22)
    doom-big-font (font-spec :family font-family :size 34))

(setq display-line-numbers-type t)

(setq fancy-splash-image "~/.doom.d/notarock.png")

(setq frame-resize-pixelwise t)

; (feebleline-mode +1)
(dimmer-configure-magit)
(dimmer-configure-org)
(dimmer-mode t)
(setq highlight-indent-guides-method 'fill)
(add-hook 'prog-mode-hook 'highlight-indent-guides-mode)

(add-hook 'prog-mode-hook (lambda ()
                             (setq show-trailing-whitespace t)))
(add-hook 'org-mode-hook (lambda ()
                             (setq show-trailing-whitespace t)))
(add-hook 'markdown-mode-hook (lambda ()
                             (setq show-trailing-whitespace t)))

(defun my/set-initial-frame ()
  "Set initial frame size and position"
  (let* ((base-factor 0.80)
         (a-width (* (display-pixel-width) (/ base-factor 2)))
         (a-height (* (display-pixel-height) base-factor))
         (a-left (truncate (/ (- (display-pixel-width) a-width) 2)))
         (a-top (truncate (/ (- (display-pixel-height) a-height) 2))))
    (set-frame-position (selected-frame) a-left a-top)
    (set-frame-size (selected-frame) (truncate a-width)  (truncate a-height) t)))

(my/set-initial-frame)

(defun indent-buffer ()
  "Indent the whole buffer"
  (interactive)
  (save-excursion
    (indent-region (point-min) (point-max) nil)))

(defun insert-random-hash ()
  (interactive)
  (insert (string-trim (shell-command-to-string "< /dev/urandom tr -dc _A-Z-a-z-0-9 | head -c${1:-64};echo;"))))

(evil-ex-define-cmd "q" 'kill-this-buffer)
(evil-ex-define-cmd "quit" 'evil-quit)

(setq lsp-python-ms-executable (executable-find "python-language-server"))

;; (add-to-list 'auto-mode-alist '("\\.ts\\'" . typescript-mode))
;; (add-to-list 'auto-mode-alist '("\\.tsx\\'" . typescript-mode))
;; (after! flycheck
  ;; (flycheck-add-mode 'javascript-eslint 'typescript-mode)
  ;; (flycheck-add-mode 'css-stylelint 'typescript-mode)
  ;; (add-hook 'typescript-mode-hook (lambda () (flycheck-add-next-checker 'lsp-ui 'javascript-eslint)))
  ;; (add-hook 'typescript-mode-hook (lambda () (flycheck-add-next-checker 'javascript-eslint 'css-stylelint))))

(add-hook 'js2-mode-hook 'prettier-js-mode)
(add-hook 'web-mode-hook 'prettier-js-mode)

(map! :map web-mode-map
      :n "SPC m F" #'eslint-fix)

(map! :map typescript-mode-map
      :n "SPC m F" #'eslint-fix)

(use-package org-fancy-priorities
  :hook (org-mode . org-fancy-priorities-mode)
  :config
  (setq org-fancy-priorities-list '("■" "■" "■")))

(after! org
  (map! :map org-mode-map
        :n "m-j" #'org-metadown
        :n "m-k" #'org-metaup)
  (setq ispell-local-dictionary "fr-toutesvariantes"
        org-directory "~/org/"
        org-journal-file (concat org-directory "journal.org")
        org-todos-file (concat org-directory "todos.org")
        org-capture-templates       (doct '(("Journal" :keys "j"
                                             :file org-journal-file
                                             :datetree t
                                             :todo nil
                                             :template ("* %^{description}"
                                                        ":properties:"
                                                        ":created: %u"
                                                        ":end:"
                                                        "%?"))
                                            ("Todos" :keys "t"
                                             :file org-todos-file
                                             :todo-state "TODO"
                                             :template ("* TODO: %^{description}"
                                                        ":properties:"
                                                        ":created: %u"
                                                        ":end:"))
                                            ))
        org-todo-keyword-faces (quote (("todo" :foreground "#ff6347" :weight bold)
                                       ("done" :foreground "#006400" :weight bold :strike-through t)))
        org-todo-keywords '((sequence "todo(t)" "done(d)"))
        org-log-done t))

(after! markdown
  (setq ispell-local-dictionary "fr-toutesvariantes"))

(map! :ne "C-S-k" #'drag-stuff-up)
(map! :ne "C-S-j" #'drag-stuff-down)
(map! :ne "C-S-l" #'drag-stuff-right)
(map! :ne "C-S-h" #'drag-stuff-left)

(map! :ne "SPC =" #'indent-buffer)
(map! :ne "SPC #" #'comment-or-uncomment-region)

(map! :ne "SPC j g" #'dumb-jump-go)
(map! :ne "SPC j b" #'dumb-jump-back)

(map! :ne "SPC i h" #'insert-random-hash)

(map! :ne "SPC w V" (lambda () (interactive)(evil-window-vsplit) (other-window 1)))

(map! [remap org-capture] nil)

;;  (shell-command "pandoc config.org -o README.md")

(global-git-gutter-mode +1)

(global-wakatime-mode +1)

(setq ispell-dictionary "en_US")
